osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs

```diff
@@ -1,23 +1,82 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Game.Configuration;
 using osu.Game.Rulesets.Mods;
 using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Taiko.Objects;
+using osu.Game.Rulesets.Taiko.Objects.Drawables;
+using osu.Game.Rulesets.Taiko.UI;
+using osu.Game.Rulesets.Timing;
+using osu.Game.Rulesets.UI;
 
 namespace osu.Game.Rulesets.Taiko.Mods
 {
-    public class TaikoModHidden : ModHidden
+    public class TaikoModHidden : ModHidden, IApplicableToDrawableRuleset<TaikoHitObject>
     {
         public override string Description => @"Beats fade out before you hit them!";
         public override double ScoreMultiplier => 1.06;
-        public override bool HasImplementation => false;
+        private const double relative_fade_position = 0.5;
+        private DrawableTaikoRuleset ruleset;
+        private readonly MultiplierControlPoint searchPoint = new MultiplierControlPoint();
+
+        [SettingSource("Hide Time", "Controls the time the note stays hidden")]
+        public BindableNumber<double> VisibilityMod { get; } = new BindableDouble
+        {
+            MinValue = 0.35,
+            // Max visibility is only to be used with max playfield size
+            MaxValue = 1.35,
+            Default = 1.0,
+            Value = 1.0,
+            Precision = 0.01,
+        };
 
         protected override void ApplyIncreasedVisibilityState(DrawableHitObject hitObject, ArmedState state)
         {
+            ApplyNormalVisibilityState(hitObject, state);
+        }
+
+        private MultiplierControlPoint controlPointAt(double time)
+        {
+            if (ruleset.ControlPoints.Count == 0)
+                return new MultiplierControlPoint(double.NegativeInfinity);
+
+            if (time < ruleset.ControlPoints[0].StartTime)
+                return ruleset.ControlPoints[0];
+
+            searchPoint.StartTime = time;
+            int index = ruleset.ControlPoints.BinarySearch(searchPoint);
+
+            if (index < 0)
+                index = ~index - 1;
+
+            return ruleset.ControlPoints[index];
```

Comment:
As far as I can see this is copied verbatim from `OverlappingScrollAlgorithm`. It should be shared somehow, possibly by subclassing `SortedList<MultiplierControlPoint>` and adding this there as an instance method.